host-controller/src/main/java/org/jboss/as/domain/controller/operations/SyncServerStateOperationHandler.java

```diff
@@ -126,30 +148,51 @@ public void handleResult(OperationContext.ResultAction resultAction, OperationCo
         }, OperationContext.Stage.MODEL);
     }
 
-    private Map<String, Boolean> determineServerStateChanges(OperationContext context, Resource domainRootResource, ServerOperationResolver resolver,
-                                                             ContentDownloader contentDownloader) {
-        final Map<String, Boolean> serverStateChanges = new HashMap<>();
+    private Map<String, SyncServerResultAction> determineServerStateChanges(final OperationContext context, final Resource domainRootResource, final ServerOperationResolver resolver,
+                                                                            final ContentDownloader contentDownloader) {
+        final Map<String, SyncServerResultAction> serverStateChanges = new HashMap<>();
         for (ModelNode operation : operations) {
-            PathAddress addr = PathAddress.pathAddress(operation.get(OP_ADDR));
+            final PathAddress addr = PathAddress.pathAddress(operation.get(OP_ADDR));
             contentDownloader.checkContent(operation, addr);
-
-            Map<Set<ServerIdentity>, ModelNode> serverMap = resolver.getServerOperations(context, operation, addr);
+            final ImmutableManagementResourceRegistration registration = context.getResourceRegistration();
+            final Map<Set<ServerIdentity>, ModelNode> serverMap = resolver.getServerOperations(context, operation, addr);
             for (Map.Entry<Set<ServerIdentity>, ModelNode> entry : serverMap.entrySet()) {
-                ModelNode op = entry.getValue();
-                boolean restart = op.get(OP).asString().equals(ServerProcessStateHandler.REQUIRE_RESTART_OPERATION);
-                for (ServerIdentity id : entry.getKey()) {
-                    String serverName = id.getServerName();
-                    Boolean existing = serverStateChanges.get(serverName);
-                    if (existing == null || (!existing && restart)) {
-                        serverStateChanges.put(serverName, restart);
+                final ModelNode op = entry.getValue();
+                final String opName = op.get(OP).asString();
```

Comment:
I don't think you can safely use this op for the registration.getAttributeAccess/registration.getOperationFlags check. It's the op that's expected to be executed on the server, which may not be the op that is registered here on the HC. Above you are using the address from the op that is being executed on the server. And rightly so as you have no idea if the server-op address is registered on the HC. For any /subsytem=* address without the preceding /profile=* level it won't be.

This is kind of a nit, since in practice it's likely going to just work, particularly since the "boolean restart = op.get(OP).asString().equals(ServerProcessStateHandler.REQUIRE_RESTART_OPERATION);" check that was at L139 will handle a lot of the odd cases. But it's the kind of thing that can might break some day in a hard to understand way.

Some possibilities:

1) Limit this second level check to addresses in the /profile=* tree where addr.size() > 1 (i.e. not the profile root address itself). For those address, ServerOperationResolver just takes the domain level op and strips the /profile=* part off the address, i.e. there won't be a change of the op name.

With the existing L139 "boolean restart = ..." check restored, this approach then just leaves a hole for domain/host ops outside the /profile=* tree where ServerOperationResolver isn't already returning ServerProcessStateHandler.REQUIRE_RESTART_OPERATION. Those cases really shouldn't exist. There are two reasons a reload is insufficient and restart is required:

1) The op changes something about how the HC would launch the server. The ServerOperationResolver logic handles these by returning  ServerProcessStateHandler.REQUIRE_RESTART_OPERATION.
2) Some lib we integrate stores data in a static field and that can't be changed without restart. We try really hard not to do this with kernel resources though, and if we do the ServerOperationResolver logic can handle it as well. So this case is really just a problem with extensions where the kernel has no idea what the extension will do.

2) If we decide 1) is wrong or too hacky, perhaps we can check for cases where the op name is different for 'op' versus 'operation'. In that case we know something unusual has happened and the lookup against the HC's registration is not a valid algorithm. So then just default to SyncServerResultAction.RESTART_REQUIRED.